File: cpp/437.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int sum) {
        if(!root){
            return 0;
        }
        return sumHelper(0, sum, root) + pathSum(root->left, sum) + pathSum(root->right, sum);
    }
    
    int sumHelper(int curVal, int sum, TreeNode*root){
        if(!root){
            return 0;
        }
        curVal += root->val;
        return (curVal==sum) + sumHelper(curVal, sum, root->left) + sumHelper(curVal, sum, root->right);
    }
};

File: golang/437/437.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}

	res := 0
	helper(root, 0, &sum, &res)

	return res + pathSum(root.Left, sum) + pathSum(root.Right, sum)
}

func helper(root *TreeNode, currentSum int, sum, res *int) {
	if root == nil {
		return
	}

	currentSum += root.Val
	if currentSum == *sum {
		*res += 1
	}

	helper(root.Left, currentSum, sum, res)
	helper(root.Right, currentSum, sum, res)
}
